File: Week04/127.单词接龙.go
/*
 * @lc app=leetcode.cn id=127 lang=golang
 *
 * [127] 单词接龙
 */

// @lc code=start
func ladderLength(beginWord string, endWord string, wordList []string) int {
    if !ContainsInArray(wordList, endWord) {
		return 0
	}

	lookup := Construct_words(wordList)
	visited := make([]string, 10)
	return BFSWord([]string{beginWord}, endWord, 1, lookup, visited)
}

func Construct_words(wordList []string) (lookup map[string][]string) {
	lookup = make(map[string][]string)
	for _, word := range wordList {
		for i := 0; i < len(word); i++ {
			mask_word := word[:i] + "*" + word[i+1:]
			_, ok := lookup[mask_word]
			if !ok {
				lookup[mask_word] = make([]string, 1)
			}
			lookup[mask_word] = append(lookup[mask_word], word)
		}
	}
	return
}

func BFSWord(beginWords []string, endWord string, level int, lookup map[string][]string, visited []string) int {
	if ContainsInArray(beginWords, endWord) {
		return level
	}

	next_words := make([]string, 0)
	for _, beginWord := range beginWords {
		for i := 0; i < len(beginWord); i++ {
			mask_word := beginWord[:i] + "*" + beginWord[i+1:]
			found_words, ok := lookup[mask_word]
			if ok {
				for _, word := range found_words {
					if !ContainsInArray(visited, word) && !ContainsInArray(beginWords, word) && !ContainsInArray(next_words, word) {
						next_words = append(next_words, word)
					}
				}
			}
		}
	}

	if len(next_words) != 0 {
		return BFSWord(next_words, endWord, level + 1, lookup, append(visited, beginWords...))
	}

	return 0
}

func ContainsInArray(wordList []string, key string) bool {
	for _, word := range wordList {
		if key == word {
			return true
		}
	}
	return false
}
// @lc code=end

File: Week04/127.单词接龙.py
#
# @lc app=leetcode.cn id=127 lang=python3
#
# [127] 单词接龙
#

# @lc code=start
from collections import defaultdict, deque
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0

        lookup = defaultdict(list)
        for word in wordList:
            for i in range(len(word)):
                mask_word = word[:i] + "*" + word[i+1:]
                lookup[mask_word].append(word)

        return self.bfs(beginWord, endWord, lookup)
    
    def bfs(self, beginWord, endWord, lookup):
        queue, visited = deque([(beginWord, 1)]), {beginWord: 1}
        while queue:
            word, steps = queue.popleft()
            if word == endWord:
                return steps
            
            for i in range(len(word)):
                mask = word[:i] + "*" + word[i+1:]
                for neigh in lookup.get(mask, []):
                    if neigh not in visited:
                        visited[neigh] = steps + 1
                        queue.append((neigh, steps + 1))
        return 0

    # def bfs(self, beginWord, endWord, lookup):
    #     left_queue, right_queue = deque([(beginWord, 1)]), deque([(endWord, 0)])
    #     left_visited = {beginWord: 1}
    #     right_visited = {endWord: 0}
    #     while left_queue and right_queue:
    #         ans = self.visitNode(lookup, left_queue, left_visited, right_visited)
    #         if ans: return ans
    #         ans = self.visitNode(lookup, right_queue, right_visited, left_visited)
    #         if ans: return ans
    #     return 0

    # def visitNode(self, lookup, queue, visited, other_visited):
    #     word, steps = queue.popleft()
    #     for i in range(len(word)):
    #         mask = word[:i] + "*" + word[i+1:]
    #         for neigh in lookup.get(mask, []):
    #             if neigh in other_visited:
    #                 return steps + 1 + other_visited[neigh]
    #             if neigh not in visited:
    #                 visited[neigh] = steps + 1
    #                 queue.append((neigh, steps + 1)) 
    #     return None
# @lc code=end

